packaging/rest-service/files/opt/cloudify/encryption/update-encryption-key

```diff
@@ -0,0 +1,129 @@
+#! /opt/manager/env/bin/python
+import argparse
+import base64
+import json
+import os
+import sys
+
+from manager_rest.flask_utils import setup_flask_app
+from manager_rest.storage import get_storage_manager, models
+from manager_rest.cryptography_utils import (
+    _get_encryption_key,
+    decrypt,
+    decrypt128,
+    encrypt,
+)
+from cryptography.fernet import InvalidToken
+
+
+DECRYPT_FUNCTIONS = {
+    32: decrypt128,
+    64: decrypt,
+}
+NEW_ENCRYPTION_KEY_STAGING_PATH = '/opt/manager/new_encryption_key'
+
+
+def update_encryption_key(old_key, new_key, commit=False):
+    decrypt_func = DECRYPT_FUNCTIONS[len(
+        base64.urlsafe_b64decode(old_key)
+    )]
+
+    failures = []
+    rmq_failures = []
+
+    with setup_flask_app().app_context():
+        sm = get_storage_manager()
+        secrets = sm.list(models.Secret)
+        for secret in secrets:
+            try:
+                val = decrypt_func(secret.value, old_key)
+            except InvalidToken:
+                try:
+                    decrypt(secret.value, new_key)
+                except InvalidToken:
+                    failures.append((secret.key, secret.tenant_name))
+                continue
+            new_val = encrypt(val, new_key)
+            if commit:
+                secret.value = new_val
+                sm.update(secret)
+
+        tenants = sm.list(models.Tenant)
+        for tenant in tenants:
+            try:
+                val = decrypt_func(tenant.rabbitmq_password, old_key)
+            except InvalidToken:
+                try:
+                    decrypt(tenant.rabbitmq_password, new_key)
+                except InvalidToken:
+                    rmq_failures.append(tenant.name)
+                continue
+            new_val = encrypt(val, new_key)
+            if commit:
```

Comment:
Would a single sm.update work?
The other problem is that our sm.update function notes that not everything needs update calling to be updated, so the value might actually change when it is changed rather than on the update being called?

Reply:
no, sm.update is unfortunately single value
don't worry about the other thing - if you just change the value, it might well be changed in the db - but only if you actually save it at some point - and here, we only save using the update